src/Stratis.Bitcoin.Tests/Consensus/ChainedHeaderTreeTest.cs

```diff
@@ -1966,6 +2030,122 @@ public void ChainWith10DataAvailableHeaders_PartialValidationSucceededOnTheFirst
             listOfHeaders.Should().HaveCount(1);
             listOfHeaders.First().ChainedHeader.HashBlock.Should().Be(secondHeader.HashBlock);
         }
+      
+        /// <summary>
+        /// Issue 19 @ New peer K claims 10d. Peer K disconnects. Chain shouldn't change.
+        /// </summary>
+        [Fact]
+        public void NewPeerClaimsHeadSecondTime_PeerDisconnected_ChainShouldNotChange()
+        {
+            const int initialChainSize = 5;
+            TestContext ctx = new TestContextBuilder().WithInitialChain(initialChainSize).UseCheckpoints().Build();
+            ChainedHeaderTree cht = ctx.ChainedHeaderTree;
+            ChainedHeader initialChainTip = ctx.InitialChainTip;
+
+            IEnumerable<BlockHeader> listOfAllUniqueBlockHeaders = ctx.SetupPeersForTest(initialChainSize, ctx, cht, initialChainTip, out ChainedHeader chainATip, out ChainedHeader chainDTip);
+            CheckChainedHeaderTreeConsistency(cht, listOfAllUniqueBlockHeaders);
+
+            //Additional SetUp for current test
+            ChainedHeader chainKTip = chainDTip; //peer K has exactly the same chain as peer D.
+            List<BlockHeader> peerKBlockHeaders = ctx.ChainedHeaderToList(chainKTip, chainKTip.Height);
+
+            //Claiming Peer K chain
+            ConnectNewHeadersResult connectionPeerKResult = cht.ConnectNewHeaders(5, peerKBlockHeaders);
+            ctx.ClaimPeerChain(cht, chainKTip, connectionPeerKResult);
+
+            Dictionary<uint256, ChainedHeader> chainedHeadersWithPeerK = cht.GetChainedHeadersByHash();
+
+            cht.PeerDisconnected(5);
+
+            Dictionary<uint256, ChainedHeader> chainedHeadersWithoutPeerK = cht.GetChainedHeadersByHash();
+
+            chainedHeadersWithoutPeerK.Should().BeEquivalentTo(chainedHeadersWithPeerK);
```

Comment:
Is this not just comparing the references?

Reply:
no, it's not just reference.
From description for this method:
 Summary:
     Asserts that two dictionaries are equivalent.
// Remarks:
        //     The values within the dictionaries are equivalent when both object graphs have
        //     equally named properties with the same value, irrespective of the type of those
        //     objects. Two properties are also equal if one type can be converted to another
        //     and the result is equal. The type of the values in the dictionaries are ignored
        //     as long as both dictionaries contain the same keys and the values for each key
        //     are structurally equivalent. Notice that actual behavior is determined by the
        //     global defaults managed by the FluentAssertions.AssertionOptions class.

Reply:
Okay, thank you.

Reply:
The method itself is not comparing the references. However `chainedHeadersWithoutPeerK` is the same dictionary as `chainedHeadersWithPeerK` so the logic is not correct. 
To avoid comparing this guy against itself use 
`var chainedHeadersWithPeerK = new Dictionary<uint256, ChainedHeader>(cht.GetChainedHeadersByHash());`

instead of 
`Dictionary<uint256, ChainedHeader> chainedHeadersWithPeerK = cht.GetChainedHeadersByHash();`